File: shared/src/main/java/ru/privatenull/pnauth/policy/AuthAccessService.java
package ru.privatenull.pnauth.policy;

import ru.privatenull.pnauth.api.AuthApi;
import ru.privatenull.pnauth.message.AuthMessages;
import ru.privatenull.pnauth.config.ProxySettings;

import java.util.Locale;
import java.util.UUID;

public final class AuthAccessService {
    private final AuthApi auth;
    private final ProxySettings proxySettings;
    private final AccessSettings accessSettings;
    private final AuthMessages messages;

    public AuthAccessService(
            AuthApi auth,
            ProxySettings proxySettings,
            AccessSettings accessSettings,
            AuthMessages messages
    ) {
        this.auth = auth;
        this.proxySettings = proxySettings;
        this.accessSettings = accessSettings;
        this.messages = messages;
    }

    public AccessDecision command(UUID uniqueId, String commandLine) {
        if (auth.isAuthenticated(uniqueId) || accessSettings.unauthenticatedCommands().contains(commandName(commandLine))) {
            return AccessDecision.ALLOW;
        }
        return AccessDecision.DENY;
    }

    public AccessDecision chat(UUID uniqueId) {
        return !accessSettings.blockChat() || auth.isAuthenticated(uniqueId)
                ? AccessDecision.ALLOW
                : AccessDecision.DENY;
    }

    public ServerAccessDecision server(UUID uniqueId, String serverName) {
        if (!proxySettings.requireServerAuth()
                || auth.isAuthenticated(uniqueId)
                || serverName.equalsIgnoreCase(proxySettings.authServer())) {
            return ServerAccessDecision.ALLOW;
        }
        return ServerAccessDecision.REDIRECT_TO_AUTH;
    }

    public String blockedMessage() {
        return messages.text("access.blocked");
    }

    public String message(String key) {
        return messages.text(key);
    }

    public String authServerMissingMessage() {
        return messages.text("access.auth_server_missing", java.util.Map.of("server", proxySettings.authServer()));
    }

    public String authServerName() {
        return proxySettings.authServer();
    }

    private static String commandName(String commandLine) {
        String command = commandLine == null ? "" : commandLine.trim().toLowerCase(Locale.ROOT);
        if (command.startsWith("/")) {
            command = command.substring(1);
        }
        int separator = command.indexOf(' ');
        return separator < 0 ? command : command.substring(0, separator);
    }

    public enum AccessDecision {
        ALLOW,
        DENY
    }

    public enum ServerAccessDecision {
        ALLOW,
        REDIRECT_TO_AUTH
    }
}

File: shared/src/main/kotlin/ru/privatenull/pnauth/policy/AuthAccessService.kt
package ru.privatenull.pnauth.policy

import ru.privatenull.pnauth.api.AuthApi
import ru.privatenull.pnauth.config.ProxySettings
import ru.privatenull.pnauth.message.AuthMessages
import java.util.Locale
import java.util.UUID

class AuthAccessService(
    private val auth: AuthApi,
    private val proxySettings: ProxySettings,
    private val accessSettings: AccessSettings,
    private val messages: AuthMessages
) {
    fun command(uniqueId: UUID, commandLine: String?): AccessDecision {
        return if (auth.isAuthenticated(uniqueId) || accessSettings.unauthenticatedCommands.contains(commandName(commandLine))) {
            AccessDecision.ALLOW
        } else {
            AccessDecision.DENY
        }
    }

    fun chat(uniqueId: UUID): AccessDecision {
        return if (!accessSettings.blockChat || auth.isAuthenticated(uniqueId)) {
            AccessDecision.ALLOW
        } else {
            AccessDecision.DENY
        }
    }

    fun server(uniqueId: UUID, serverName: String): ServerAccessDecision {
        return if (!proxySettings.requireServerAuth || auth.isAuthenticated(uniqueId) || proxySettings.isAuthServer(serverName)) {
            ServerAccessDecision.ALLOW
        } else {
            ServerAccessDecision.REDIRECT_TO_AUTH
        }
    }

    fun blockedMessage(): String {
        return messages.text("access.blocked")
    }

    fun message(key: String): String {
        return messages.text(key)
    }

    fun authServerMissingMessage(): String {
        return messages.text(
            "access.auth_server_missing",
            mapOf("server" to proxySettings.authServers.joinToString(", "))
        )
    }

    fun authServerName(): String = proxySettings.authServer

    fun authServerNames(): List<String> = proxySettings.authServers

    fun isAuthServer(serverName: String?): Boolean = proxySettings.isAuthServer(serverName)

    enum class AccessDecision {
        ALLOW,
        DENY
    }

    enum class ServerAccessDecision {
        ALLOW,
        REDIRECT_TO_AUTH
    }

    companion object {
        private fun commandName(commandLine: String?): String {
            var command = commandLine?.trim()?.lowercase(Locale.ROOT) ?: ""
            if (command.startsWith("/")) {
                command = command.substring(1)
            }
            val separator = command.indexOf(' ')
            return if (separator < 0) command else command.substring(0, separator)
        }
    }
}
